tamagotchi/io/src/lib.rs

```diff
@@ -13,20 +15,162 @@ impl Metadata for TamagotchiMetadata{
+    pub entertained_block: u64,
+    pub rested: u64,
+    pub rested_block: u64,
+    pub allowed_account: Option<ActorId>,
+    pub ft_contract: Option<ActorId>,
+    pub transaction_id: u64,
+    pub approve_transaction: Option<(TransactionId, ActorId, u128)>,
+}
+
+impl TamagotchiState {
+    pub fn feed(&mut self, current_block_height: u64) {
+        debug!("Fed block {:?}", self.fed_block);
+        debug!("Total Height {:?}", current_block_height - self.fed_block);
+        self.fed_block = current_block_height;
+        self.fed += FILL_PER_FEED
+            .saturating_sub(HUNGER_PER_BLOCK * (current_block_height - self.fed_block));
+        TamagotchiState::verify_limit(&mut self.fed);
+    }
+
+    pub fn play(&mut self, current_block_height: u64) {
+        self.entertained_block = current_block_height;
+        self.entertained += FILL_PER_ENTERTAINMENT
+            .saturating_sub(BOREDOM_PER_BLOCK * (current_block_height - self.entertained_block));
+        TamagotchiState::verify_limit(&mut self.entertained);
+    }
+
+    pub fn sleep(&mut self, current_block_height: u64) {
+        TamagotchiState::verify_limit(&mut self.rested);
+        self.rested_block = current_block_height;
+        self.rested += FILL_PER_SLEEP
+            .saturating_sub(ENERGY_PER_BLOCK * (current_block_height - self.rested_block));
+        TamagotchiState::verify_limit(&mut self.rested);
+    }
+
+    pub fn verify_limit(mood_param: &mut u64) {
+        if *mood_param > MAX_MOOD_VALUE {
+            *mood_param = MAX_MOOD_VALUE;
+        }
+        if *mood_param < MIN_MOOD_VALUE {
+            *mood_param = MIN_MOOD_VALUE;
+        }
+    }
+
+    pub fn verify_ownership(&self, source: ActorId) -> bool {
+        self.owner == source
+    }
+
+    pub fn verify_allowed_account(&self, source: ActorId) -> bool {
+        match self.allowed_account {
+            Some(account) => account == source,
+            None => false,
+        }
+    }
+
+    pub fn verify_permission(&self, source: ActorId) -> bool {
+        self.verify_ownership(source) || self.verify_allowed_account(source)
+    }
+
+    pub async fn approve_tokens(&mut self, account: &ActorId, amount: u128) -> TmEvent {
+        let (transaction_id, account, amount) = if let Some((
+            ft_transaction_id,
+            prev_account,
+            prev_amount,
+        )) = self.approve_transaction
+        {
+            if prev_account != *account || prev_amount != amount {
+                panic!("Please complete the previous transaction");
+            } else {
+                (ft_transaction_id, prev_account, prev_amount)
+            }
+        } else {
+            let ft_transaction_id = self.transaction_id;
+            self.transaction_id = self.transaction_id.wrapping_add(1);
+            self.approve_transaction = Some((ft_transaction_id, *account, amount));
+            (ft_transaction_id, *account, amount)
+        };
+        if let Some(contract) = self.ft_contract {
```

Comment:
`if let Some(a) = b { ... } else { panic!("Oops") }` can be replaced with `a = b.expect("Oops"); ...`